### DISCUSSION TITLE
Using custom data with the classification demo and color camera module

### DISCUSSION THREAD

**Discussion by ecselab2** [bitcraze/discussions#697] [Q&A]
Hey there,

I have a couple of problems with the classification demo. I want to train a model to see packages with different colors.

1. You can not save pictures directly in the required JEPG format (not sure if it really has to be JPEG but the demo says so) with the opencv-viewer.py saving is only available for raw images.
2. When I add custom data the generated model still only recognizes the background and package when I open the cfclient like the classification demo eventhough more classes have been added.  I adjusted the final dense layer according to the screenshot below.
![grafik](https://user-images.githubusercontent.com/132353639/235650064-8985ff64-fcf8-4f2c-8363-ff8477bdd455.png)
3. I am not sure if I should take the colored raw pictures and format them into the .jpeg format (this is what I did) or if I should approach this differently.

Thanks for the help <3

**Comment by knmcguire** [comment]
Hi!

You should probably also change the classification.c in the example as well as that is now hardcoded to only have 2 outputs: 

https://github.com/bitcraze/aideck-gap8-examples/blob/8367be931fc041f48843de7f3ba8ecc250cd3469/examples/ai/classification/classification.c#L79-L86

about the format, I'm not sure how to help with that but  you can consider using a different library for the conversion but you can also check what Keras accepts as training input. In the end it all comes down to intensities per pixel anyway.

**Comment by ecselab2** [comment]
@knmcguire thank you for your answer <3. I adjusted the classification.c file like this.


![important](https://github.com/bitcraze/discussions/assets/132353639/46e6d52a-985b-4c78-98ca-baa646e83e55)
1. My problem now is that it still only sees one packet the raspy packet (named so because i took a red raspberry pi packaging as a packet). And occasionally orange gets recognized. But the raspberry package always non stop. 
2. Is there something I have to change and do I have to add the background as a class or not because I want to filter out different colors and not shapes. 
3. What exactly is Output_1[0] Output_1[1] and Output_1[3]. I get that they represent the confidence that the model has that it identified a certain package. But how do I map these to the right names.

For example I have the following folders in the training data folder.
![grafik](https://github.com/bitcraze/discussions/assets/132353639/395c6a28-fa8b-42a6-906f-6651fe3d6432)
would Output_1[0] be white Output_1[1] be raspy and Output_1[2] be orange. I looked at the code that you sent and the background is also listed as Output_1[1] although it is alphabetically first.

Thank you for the help

**Comment by knmcguire** [comment]
That depends on how Keras organizes the classes, which I'll need to google myself how it does that. Like here: 

https://datascience.stackexchange.com/questions/28539/how-can-i-find-out-what-class-each-of-the-columns-in-the-probabilities-output-co
